common/src/enforcer/source/cargo/mod.rs

```diff
@@ -0,0 +1,55 @@
+use crate::enforcer::{source::Source, Dependency};
+use crate::highlight::Range;
+use anyhow::anyhow;
+use async_trait::async_trait;
+use std::path::PathBuf;
+use url::Url;
+
+fn cargo_to_purl(name: String, dep: cargo_toml::Dependency) -> Option<Dependency> {
+    match dep {
+        cargo_toml::Dependency::Simple(version) => Some(Dependency {
+            purl: Url::parse(format!("pkg:cargo/{name}@{version}").as_str()).unwrap(),
+        }),
+        cargo_toml::Dependency::Detailed(detail) => {
+            // if there is no version provided we can't build a compliant package url
+            detail.version.map(|version| Dependency {
+                purl: Url::parse(format!("pkg:cargo/{}@{}", name, version).as_str()).unwrap(),
+            })
+        }
+        cargo_toml::Dependency::Inherited(_) => unimplemented!(),
+    }
+}
+
+pub struct CargoSource {
+    root: PathBuf,
+}
+
+impl CargoSource {
+    pub fn new(root: impl Into<PathBuf>) -> Self {
+        Self { root: root.into() }
+    }
+}
+
+#[async_trait]
+impl Source for CargoSource {
+    async fn scan(&self) -> anyhow::Result<Vec<Dependency>> {
+        let manifest = cargo_toml::Manifest::from_path(self.root.join("Cargo.toml"))?;
+
+        // let content = fs::read_to_string(self.root.join("Cargo.toml"))?;
+        // let manifest = cargo_toml::Manifest::from_str(&content)?;
+
+        Ok(manifest
```

Comment:
To my understanding, this is only finding the declared dependencies. However, we should find all (including the transient ones).

Therefore I should leverage existing tooling, which already does that for us. I think this could be done using the functionality from the `cargo cyclonedx` tool. We might "shell out" to generate this, as using Rust ourselves, we might be able to embed this (which would not require anyone to install additional tooling!).

Alternatively, I think we might be able to process the "lock" file instead. Which would require less dependencies, but still give a more accurate view on things.